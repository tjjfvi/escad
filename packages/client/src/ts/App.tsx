
import React from "react";
import Disconnected from "./Disconnected";
import Preview from "./Preview";

const App = () => <>
  <Preview/>
  <Disconnected/>
</>

export default App;
