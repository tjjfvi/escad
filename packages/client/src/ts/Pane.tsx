// @ts-nocheck

import React from "react";
import { useObservable } from "rhobo";

const Pane = ({ name, className = name, children, right = false, left = !right }) => {
  let open = useObservable.use(false);
  return <div className={(open() ? "open " : "") + (left ? "left" : "right") + " Pane " + className}>
    <div className="side" onClick={() => open.toggle()}><span>{name}</span></div>
    <div>{children}</div>
  </div>
}

export default Pane;
