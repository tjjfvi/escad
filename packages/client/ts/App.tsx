
import React from "react";
import Disconnected from "./Disconnected";
import Preview from "./Preview";
import Parameters from "./Parameters";
import Hierarchy from "./Hierarchy";

const App = () => <>
  <Parameters />
  <Preview />
  <Hierarchy />
  <Disconnected />
</>

export default App;
