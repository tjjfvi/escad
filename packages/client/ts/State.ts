/* @flow */

import { observable } from "rhobo";
import WS from "./ws";

class State {

  ws = new WS("ws" + window.location.toString().slice(4) + "ws/", this);

  connected = observable<boolean>(false);
  id = observable<string>();
  serverId = observable<string>();
  paramDef = observable<any>([]);
  params = observable<any>({});
  hierarchy = observable<any>();
  shas = observable<Array<string>>([]);

  constructor() {
    // this.ws.on("hierarchy", d => this.hierarchy(this.processHierarchy(d)));
    this.ws.on("shas", d => this.shas(d));
    this.ws.on("paramDef", d => this.paramDef(d));
    this.params.on("update", () => this.ws.s("params", this.params()))
  }

  // processHierarchy({ name, important, shas, children }) {
  //   children = children.map(x => this.processHierarchy(x));
  //   shas = shas.length ? shas : children.flatMap(x => (x.pre[0] || x).shas);
  //   let tree = { children, shas, name, important, pre: [], post: [], meta: {} }
  //   tree.leaf = children.every(c => c.leaf && !c.important);
  //   let child = children.length === 1 ? children[0] : null;
  //   Object.assign(tree,
  //     !tree.leaf && child ?
  //       !child.important ?
  //         { children: child.children, post: [{ name: child.name, shas: child.shas }, ...child.post] } :
  //         important ?
  //           {} :
  //           { ...child, pre: [{ name, shas }, ...child.pre] } :
  //       {}
  //   );
  //   return tree;
  // }

}

export default new State();
export { State };
