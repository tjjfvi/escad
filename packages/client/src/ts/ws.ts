
import { EventEmitter } from "tsee";
import { State } from "./State";
import flatted from "flatted";

class WS extends EventEmitter<any> {

  state: State;
  url: string;
  ws: WebSocket;

  timeout: any;

  constructor(url: string, state: State){
    super();
    this.url = url;
    this.state = state;
    this.ws = this.initWs();
  }

  s(...data: any){
    if(this.ws.readyState !== 1)
      return;
    console.log("→", ...data);
    this.ws.send(flatted.stringify(data));
  }

  initWs(){
    this.timeout = null;

    if(this.ws)
      this.ws.close();

    this.ws = new WebSocket(this.url);
    this.ws.addEventListener("open", () => {
      this.s("init", this.state.id(), this.state.serverId());
    });

    this.ws.addEventListener("error", () => {
      if(!this.timeout) this.timeout = setTimeout(() => this.initWs(), 5000);
    })

    this.ws.addEventListener("message", msg => {
      let [type, ...data] = flatted.parse(msg.data);

      this.state.connected(true);
      clearTimeout(this.timeout);

      this.timeout = setTimeout(() => {
        this.state.connected(false);
        this.ws.close();
        this.initWs();
      }, 5000);

      if(type === "ping")
        return;

      if(type === "id")
        this.state.id(data[0]);

      if(type === "serverId")
        this.state.serverId(data[0]);

      this.emit("message", type, ...data);
      this.emit(type, ...data);

      console.log("←", type, ...data);
    })

    return this.ws;
  }

}

export default WS;
