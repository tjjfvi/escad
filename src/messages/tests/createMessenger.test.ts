import { assertEquals, mock, snapshot } from "../../testUtils/mod.ts";
import {
  Connection,
  createConnectionPair,
  createMessenger,
  Messenger,
} from "../mod.ts";

type TestShape = {
  promiseResolve: <T>(value: T) => Promise<T>;
  bouncingPromiseResolve: <T>(value: T, count: number) => Promise<T>;
};

type TestEvents = {
  promiseResolveBounced: [value: unknown, count: number];
};

type TestMessenger = Messenger<TestShape, TestShape, TestEvents>;

const createTestMessenger = (connection: Connection<unknown>) => {
  const messenger: TestMessenger = createMessenger({
    impl: {
      promiseResolve: (value) => Promise.resolve(value),
      async bouncingPromiseResolve(value, count) {
        if (count === 0) {
          return value;
        }
        messenger.emit("promiseResolveBounced", value, count);
        return messenger.bouncingPromiseResolve(value, count - 1);
      },
    },
    connection,
  });
  return messenger;
};

Deno.test("createMessenger", async () => {
  const [a, b] = createConnectionPair();
  const fn = mock();
  a.onMsg((x) => fn("b->a", x));
  b.onMsg((x) => fn("a->b", x));
  const msgrA = createTestMessenger(a);
  const msgrB = createTestMessenger(b);
  assertEquals(await msgrA.promiseResolve(-1), -1);
  assertEquals(await msgrA.promiseResolve(0), 0);
  assertEquals(await msgrB.promiseResolve(1), 1);
  const eventFn1 = mock();
  const eventFn2 = mock();
  const eventFn3 = mock();
  const eventFn4 = mock();
  msgrA.on("promiseResolveBounced", eventFn1);
  msgrB.on("promiseResolveBounced", eventFn2);
  msgrB.once("promiseResolveBounced", eventFn3);
  (async () => {
    for await (const event of msgrA.on("promiseResolveBounced")) {
      eventFn4(event);
    }
  })();
  let p = msgrA.once("promiseResolveBounced").then((x) =>
    assertEquals(x, [
      2,
      10,
    ])
  );
  assertEquals(await msgrA.bouncingPromiseResolve(2, 10), 2);
  await p;
  assertEquals(eventFn1.calls.length, 10);
  assertEquals(eventFn2.calls.length, 10);
  assertEquals(eventFn3.calls.length, 1);
  assertEquals(eventFn4.calls.length, 10);
  b.send(1234);
  b.send(null);
  b.send({});
  b.send([]);
  b.send(["missingno", 1, 2, 3]);
  b.send(["call", null, 1, 1]);
  b.send(["call", 1, {}, 1]);
  await snapshot(import.meta.url, "", fn.calls);
});
