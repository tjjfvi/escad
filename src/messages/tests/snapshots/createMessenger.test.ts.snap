[
  [
    "a->b",
    [
      "retry",
    ],
  ],
  [
    "b->a",
    [
      "retry",
    ],
  ],
  [
    "a->b",
    [
      "call",
      1,
      "promiseResolve",
      -1,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      1,
      null,
      -1,
    ],
  ],
  [
    "a->b",
    [
      "call",
      2,
      "promiseResolve",
      0,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      2,
      null,
      0,
    ],
  ],
  [
    "b->a",
    [
      "call",
      1,
      "promiseResolve",
      1,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      1,
      null,
      1,
    ],
  ],
  [
    "a->b",
    [
      "call",
      3,
      "bouncingPromiseResolve",
      2,
      10,
    ],
  ],
  [
    "b->a",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      10,
    ],
  ],
  [
    "b->a",
    [
      "call",
      2,
      "bouncingPromiseResolve",
      2,
      9,
    ],
  ],
  [
    "a->b",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      9,
    ],
  ],
  [
    "a->b",
    [
      "call",
      4,
      "bouncingPromiseResolve",
      2,
      8,
    ],
  ],
  [
    "b->a",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      8,
    ],
  ],
  [
    "b->a",
    [
      "call",
      3,
      "bouncingPromiseResolve",
      2,
      7,
    ],
  ],
  [
    "a->b",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      7,
    ],
  ],
  [
    "a->b",
    [
      "call",
      5,
      "bouncingPromiseResolve",
      2,
      6,
    ],
  ],
  [
    "b->a",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      6,
    ],
  ],
  [
    "b->a",
    [
      "call",
      4,
      "bouncingPromiseResolve",
      2,
      5,
    ],
  ],
  [
    "a->b",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      5,
    ],
  ],
  [
    "a->b",
    [
      "call",
      6,
      "bouncingPromiseResolve",
      2,
      4,
    ],
  ],
  [
    "b->a",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      4,
    ],
  ],
  [
    "b->a",
    [
      "call",
      5,
      "bouncingPromiseResolve",
      2,
      3,
    ],
  ],
  [
    "a->b",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      3,
    ],
  ],
  [
    "a->b",
    [
      "call",
      7,
      "bouncingPromiseResolve",
      2,
      2,
    ],
  ],
  [
    "b->a",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      2,
    ],
  ],
  [
    "b->a",
    [
      "call",
      6,
      "bouncingPromiseResolve",
      2,
      1,
    ],
  ],
  [
    "a->b",
    [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      1,
    ],
  ],
  [
    "a->b",
    [
      "call",
      8,
      "bouncingPromiseResolve",
      2,
      0,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      8,
      null,
      2,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      6,
      null,
      2,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      7,
      null,
      2,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      5,
      null,
      2,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      6,
      null,
      2,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      4,
      null,
      2,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      5,
      null,
      2,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      3,
      null,
      2,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      4,
      null,
      2,
    ],
  ],
  [
    "a->b",
    [
      "resolve",
      2,
      null,
      2,
    ],
  ],
  [
    "b->a",
    [
      "resolve",
      3,
      null,
      2,
    ],
  ],
  [
    "b->a",
    1234,
  ],
  [
    "b->a",
    null,
  ],
  [
    "b->a",
    {},
  ],
  [
    "b->a",
    [
      ,
    ],
  ],
  [
    "b->a",
    [
      "missingno",
      1,
      2,
      3,
    ],
  ],
  [
    "b->a",
    [
      "call",
      null,
      1,
      1,
    ],
  ],
  [
    "b->a",
    [
      "call",
      1,
      {},
      1,
    ],
  ],
]